### PR TITLE
Add virtual tables panel

### LINKED ISSUE #275: Add a panel for VTables (avj)
Use the binary https://github.com/radare/radare2-regressions/blob/master/bins/elf/analysis/elf-virtualtable?raw=true as an example.

After `aaa ; av`:

```
Vtable Found at 0x00400ce8
0x00400ce8 : sym.A::greet
0x00400cf0 : sym.C::printValue


Vtable Found at 0x00400d08
0x00400d08 : sym.A::greet
0x00400d10 : sym.B::printValue


Vtable Found at 0x00400d28
0x00400d28 : sym.A::greet
0x00400d30 : sym.A::printValue
```

Json available with `avj` too

### PR DESCRIPTION
**Pull request by Abogical**
Fixes #275 

`avj` doesn't provide the method's name, so I had to improvise. I'm not sure if this is intentional. Nevertheless I'll be waiting for feedback.

### CODE REVIEW

File: src/cutter.cpp [hunk 1]
```diff
@@ -1253,6 +1253,36 @@ QList<ResourcesDescription> CutterCore::getAllResources()
...
+        QJsonObject vTableObject = value.toObject();
+
+        VTableDescription res;
+        res.addr = vTableObject["offset"].toVariant().toULongLong();
+        RVA methodCount = vTableObject["methods"].toVariant().toULongLong();
+
+        for(RVA methodPointer = res.addr; methodPointer < methodCount*wordSize + res.addr; methodPointer+=wordSize)
+        {
+            ClassMethodDescription method;
+            r_io_read_i(core_->io, methodPointer, &method.addr, wordSize, false); //XXX
```
Comment (xarkes) [inline src/cutter.cpp]: Maybe the best option here would have been to modify the `avj` command in radare2. I did not check if it was easier or actually better.

File: src/widgets/VTablesWidget.cpp [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+int VTableModel::columnCount(const QModelIndex &) const
+{
+    return Columns::COUNT;
+}
+
+QVariant VTableModel::data(const QModelIndex &index, int role) const
+{
+    QModelIndex parent = index.parent();
+    if(parent.isValid()){
```
Comment (xarkes) [inline src/widgets/VTablesWidget.cpp]:
```
if (parent.isValid())
{
```

File: src/widgets/VTablesWidget.cpp [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        {
+        case Qt::DisplayRole:
+            switch(index.column())
+            {
+            case NAME:
+                return res.name;
+            case ADDRESS:
+                return RAddressString(res.addr);
+            default:
+                ;
```
Comment (xarkes) [inline src/widgets/VTablesWidget.cpp]: Use break; rather than en empty line.

File: src/widgets/VTablesWidget.cpp [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+            {
+            case NAME:
+                return tr("VTable ") + QString::number(index.row() + 1);
+            case ADDRESS:
+                return RAddressString(res.addr);
+            default:
+                ;
+            }
+        default:
+            ;
```
Comment (xarkes) [inline src/widgets/VTablesWidget.cpp]: Same for all those `;`.

File: src/widgets/VTablesWidget.cpp [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+                return res.name;
+            case ADDRESS:
+                return RAddressString(res.addr);
+            default:
+                ;
+            }
+        default:
+            ;
+        }
+    }else{
```
Comment (xarkes) [inline src/widgets/VTablesWidget.cpp]:
```
}
else
{
```

**Comment by Abogical** [comment]
Waiting on radare/radare2#9513.

**Comment by xarkes** [comment]
Radare2 PR was merged.

**Comment by Abogical** [comment]
Done!

**Comment by xarkes** [comment]
Thanks, I'll manually update radare2 submodule before merging.

**Comment by Maijin** [comment]
We need to strike it from http://rada.re/gsoc/2018/tasks.html#title_11 too

File: src/cutter.cpp [hunk 6]
```diff
@@ -1253,6 +1253,36 @@ QList<ResourcesDescription> CutterCore::getAllResources()
...
+        res.addr = vTableObject["offset"].toVariant().toULongLong();
+        QJsonArray methodArray = vTableObject["methods"].toArray();
+
+        for(QJsonValueRef methodValue : methodArray)
+        {
+            QJsonObject methodObject = methodValue.toObject();
+
+            ClassMethodDescription method;
+            method.addr = methodObject["offset"].toVariant().toULongLong();
+            method.name = methodObject["name"].toString();
```
Comment (Abogical) [inline src/cutter.cpp]: Almost forgot to include this change in the previous commit. Sorry about that.

**Pull request merged**